cl/recap/mergers.py

```diff
@@ -573,29 +573,56 @@ def merge_unnumbered_docket_entries(
 
     # Look for docket entries that match by equal long description or if the
     # long description is not set.
-    matched_entries = [
-        de.pk
-        for de in des
-        if de.description == docket_entry["description"] or not de.description
-    ]
-    if matched_entries:
+    matched_entries_queryset = des.filter(
+        Q(description=docket_entry["description"]) | Q(description="")
+    )
+    if await matched_entries_queryset.aexists():
         # Return the entry that matches the long description and remove the
         # rest if there are any duplicates.
-        matched_entries_queryset = des.filter(pk__in=matched_entries)
-        winner = matched_entries_queryset.earliest("date_created")
-        matched_entries_queryset.exclude(pk=winner.pk).delete()
+        winner = await matched_entries_queryset.aearliest("date_created")
+        await matched_entries_queryset.exclude(pk=winner.pk).adelete()
         return winner
 
     # No duplicates found by long description, choose the earliest as the
     # winner; delete the rest
-    winner = des.earliest("date_created")
-    des.exclude(pk=winner.pk).delete()
+    winner = await des.aearliest("date_created")
+    await des.exclude(pk=winner.pk).adelete()
     return winner
 
 
-def get_or_make_docket_entry(
+@sync_to_async
+def add_create_docket_entry_transaction(d, docket_entry):
+    with transaction.atomic():
+        try:
+            de, de_created = (
+                DocketEntry.objects.select_related("docket")
+                .select_for_update(of=("self", "docket"))
+                .get_or_create(
+                    docket=d, entry_number=docket_entry["document_number"]
+                )
```

Comment:
It seems that doing the `select_for_update` here does not prevent other processes from reaching `get_or_create`, so the race condition is still possible.

```suggestion
               DocketEntry.objects.get_or_create(
                    docket=d, entry_number=docket_entry["document_number"]
                )
```

The approach that successfully avoids the race condition involves creating a database lock on the docket row first, as mentioned in the previous comment. Then, here, we could only use `get_or_create`.

Reply:
applied as suggested